subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/downstream/test/TestDownstreamPlugin.java

```diff
@@ -22,12 +35,36 @@
 
     public void apply(final Project project) {
         IncubatingWarning.warn("downstream-testing plugin");
-        TaskMaker.task(project, TEST_DOWNSTREAM_TASK, TestDownstreamTask.class, new Action<TestDownstreamTask>() {
+
+        project.getPlugins().apply(UploadGistsPlugin.class);
+
+        TestDownstreamTask testDownstreamTask = TaskMaker.task(project, TEST_DOWNSTREAM_TASK, TestDownstreamTask.class, new Action<TestDownstreamTask>() {
             @Override
             public void execute(TestDownstreamTask task) {
                 task.setDescription("Runs all downstream tests.");
             }
         });
+
+        final UploadGistsTask uploadGistsTask = (UploadGistsTask) project.getTasks().findByName(UploadGistsPlugin.UPLOAD_GISTS_TASK);
+
+        uploadGistsTask.setFilesToUpload((ConfigurableFileTree)
+            project
+                .fileTree(testDownstreamTask.getLogDirectory())
+                .include("*.log")
+        );
+
+        // it has to be deferred because TestDownstreamReleaseTask are added by user after applying this plugin
+        DeferredConfiguration.deferredConfiguration(project, new Runnable() {
+            @Override
+            public void run() {
```

Comment:
What do you think about simplifying it? 

- we create uploadGists task before testDownstream task
- we pass an instance of uploadGists to testDownstream task
- each time the user calls addRepository method, we automatically add 'finalizedBy(uploadGists)'

This way we don't need deferred configuration block and we also don't need the marker interface.